Various performance boosts in nests/graphql

roypeled opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?
When bootstrapping a gql application, having many schemas results in a very large memory footprint.
For our project with 200 schemas, we couldn't run the server because it consumed past 8GB memory and died on OOM.

Issue Number: N/A


## What is the new behavior?
Reducing the amount of redundant compilations of the GQL data object model (metadata wise) is now possible because for the switch to map based metadata storage. Removing the redundant metadata assignments produces a tremendous reduction of memory usage (from 8GB+ to 700MB) 

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

rezoled commented:
Hey @kamilmysliwiec, I've made some improvements that drastically reduce the memory footprint of a GQL server. Could you please verify and merge this?

roypeled commented:
Hey @kamilmysliwiec, I've made some improvements that drastically reduce the memory footprint of a GQL server. Could you please verify and merge this?

roypeled commented:
1) getObjectOrInterfaceTypeNameIfExists - This function was not fixed on the move to map based metadata storage. Now instead of looking inside an entire metadata array, it fetches the item from the map

2) MetadataCollectionModel - I noticed that the interfaces list recieves duplicate instances of the same interface, so I switched the implementation to a map to remvoe duplicates.

3) LazyMetadataStorageHost.load - In very large applications (ie several millions of metadata object) we encountered an invalid arugments size thrown from V8, this is because the spread operator behind the scenes translates the values to multiple arguments. Switching to flat() solves this issue.

4) GraphQLSchemaFactory.create & LazyMetadataStorageHost.updateStorage - We noticed that the entire GQL metadata model was being rebuilt several times during the bootstrap of the application, I assume this was done because of the need to clear and rebuild the arrays every time. Now when we use maps we can't get duplicates in the metadata store, so I verify that the updateStorage callback is called only once, and I removed the clear call from the GraphQLSchemaFactory.create(). This results in a tremendous memory consumption boost! (in our case, from 8GB to 700MB)

kamilmysliwiec commented on packages/graphql/lib/schema-builder/graphql-schema.factory.ts:
```diff
@@ -60,7 +60,6 @@ export class GraphQLSchemaFactory {
       options = scalarsOrOptions;
     }
 
-    TypeMetadataStorage.clear();
```
Why this line is no longer required?

rezoled replied:
Because we save the metadata in maps. 
What I think happened was that when we used array to save the metadata, and called updateStorage in lazy-metadata.storage multiple times, the metadata would get duplicated in the storage.

For performance optimization, I removed redundant updateStorage calls, but now calling GraphQLSchemaFactory.create called for TypeMetadataStorage.clear (it expected the the metadata would be collected again via updateStorage), and the metadata would not get re-collected (because I removed duplicate updateStorage calls).

So I removed the clear call. 

Why this works?
1) Since TypeMetadataStorage has changed to maps, there is no chance of getting duplicate metadata values. 
2) When we have no duplicates, there is no need to reset and re-collect the metadata.

I hope my explanation is clear.

rezoled replied:
The only thing that bothers me, is that TypeMetadataStorage is a singleton that is generated on app bootstrap, and is not bounded by a NestJS instance

kamilmysliwiec commented on packages/graphql/lib/decorators/resolver.decorator.ts:
```diff
@@ -20,10 +20,8 @@ function getObjectOrInterfaceTypeNameIfExists(
+    TypeMetadataStorage.getObjectTypeMetadataByTarget(ctor);
   if (!objectMetadata) {
     const interfaceTypesMetadata = TypeMetadataStorage.getInterfacesMetadata();
     const interfaceMetadata = interfaceTypesMetadata.find(
```
Shouldn't we use `getInterfaceMetadataByTarget()` here as well?

rezoled replied:
Definitley. Fixing

kamilmysliwiec commented on packages/graphql/lib/schema-builder/collections/metadata-by-target.collection.ts:
```diff
@@ -1,10 +1,11 @@
     argumentType: [],
-    interface: [],
+    interface: new Map<Function, ClassMetadata>(),
     inputType: [],
     objectType: [],
     resolver: [],
```
Do you think we should refactor other collections too? (especially objectType?)

rezoled replied:
I agree, but I'm afraid of introducing a lot of changes. Some tests rely on a specific order (arguments for example), moving it to map would break the order.
I think that maybe there's a reason to rethink how the metadata storage works? Maybe a complete overhaul to have it much more optimized?

kamilmysliwiec commented:
![image](https://user-images.githubusercontent.com/23244943/192725243-c5b405c3-9d30-42b7-a18d-65267f54c478.png)

Build is broken now @roypeled
